### PR TITLE
[Feature] Get incumbent config and results

### LINKED ISSUE #150: Return the best configuration found by the optimizer
the [`get_incumbent_config`](https://github.com/automl/Auto-PyTorch/blob/db70a436cc8044a7bc03ad8846012bb3651eaca5/autoPyTorch/api/base_task.py#L1240) and `get_incumbent_results` is currently not doing anything. 

Ideally `get_incumbent_config` should search through the run history object and find the best configurations from there and return it. We could also have a flag to enable searching for results from the traditional classifiers. Also, returning the fitted pipeline as an object but this is optional.

### PR DESCRIPTION
**Pull request by ravinkohli**
Fixes #150 

Changes made-
1. removed `get_incumbent_config` from base task as it would be redundant to have two separate functions for getting configuration and the results
2. `get_incumbent_results` goes through the run history and gets the config that has the minimum cost on the validation set.
3. Also returns the additional info which includes optimisation loss on the other metrics specified  and the train loss.
4. Adds test to check if the function is working properly

### CODE REVIEW

File: autoPyTorch/api/base_task.py [hunk 1]
```diff
@@ -1245,17 +1239,32 @@ def __del__(self) -> None:
...
+        Returns:
+
+        """
+        assert self.run_history is not None, "No Run History found, search has not been called."
+        assert not self.run_history.empty(), "Run History is empty."
+
+        run_history_data = self.run_history.data
+        if not include_traditional:
+            run_history_data = dict(
+                filter(lambda elem: elem[1].additional_info is not None and 'trainer_configuration' not in elem[1].
```
Comment (ArlindKadra) [inline autoPyTorch/api/base_task.py:1260]: Hmm, what does having the key `trainer_configuration `mean in this scenario ? Since when include_traditional is activated, we do not filter the `run_history_data`.

File: autoPyTorch/api/base_task.py [hunk 2]
```diff
@@ -1245,17 +1239,32 @@ def __del__(self) -> None:
         # until the estimator is deleted
         self._backend.context.delete_directories(force=False)
 
-    @typing.no_type_check
     def get_incumbent_results(
-            self
-    ):
-        pass
```
[the remark this answers is not in the document]
Comment (ravinkohli) [reply autoPyTorch/api/base_task.py:1252]: they were not. We could get them by taking the last config in incumbent_trajectory or by manually going through the run history of the estimator

File: autoPyTorch/api/base_task.py (hunk 1 above)
Comment (ravinkohli) [reply autoPyTorch/api/base_task.py:1260]: so this is an ad-hoc way of checking if the run history belongs to the traditional pipeline or an autopytorch pipeline. I'll add a comment before this to make it clearer

Comment (ArlindKadra) [reply autoPyTorch/api/base_task.py:1260]: Aha, oki, perfect.

File: autoPyTorch/api/base_task.py [hunk 3]
```diff
@@ -657,9 +649,11 @@ def _do_traditional_prediction(self, num_run: int, time_left: int, func_eval_tim
                             f"Fitting {cls} took {runtime}s, performance:{cost}/{additional_info}")
                         configuration = additional_info['pipeline_configuration']
                         origin = additional_info['configuration_origin']
+                        additional_info.pop('pipeline_configuration')
```
Comment (franchuterivera) [inline autoPyTorch/api/base_task.py:652]: Sorry, why are you doing pop here?

File: autoPyTorch/api/base_task.py [hunk 4]
```diff
@@ -1245,17 +1239,32 @@ def __del__(self) -> None:
...
-    @typing.no_type_check
-    def get_incumbent_config(
-            self
-    ):
-        pass
+        Returns:
+
+        """
+        assert self.run_history is not None, "No Run History found, search has not been called."
+        assert not self.run_history.empty(), "Run History is empty."
```
Comment (franchuterivera) [inline autoPyTorch/api/base_task.py:1255]: I think this makes more sense as a ValueError? so that users can catch it if they want. Probably with a message saying:
```
Something went wrong, smac was not able to fit any model?
```

File: autoPyTorch/api/base_task.py [hunk 5]
```diff
@@ -1245,17 +1239,32 @@ def __del__(self) -> None:
...
         self._backend.context.delete_directories(force=False)
 
-    @typing.no_type_check
     def get_incumbent_results(
-            self
-    ):
-        pass
+        self,
+        include_traditional: bool = False
+    ) -> Tuple[Configuration, Dict]:
```
Comment (franchuterivera) [inline autoPyTorch/api/base_task.py:1245]: This should be a Dict[str, Union[str, int, float]]?

File: test/test_api/test_api.py [hunk 6]
```diff
@@ -177,6 +179,13 @@ def test_tabular_classification(openml_id, resampling_strategy, backend):
     score = estimator.score(y_pred, y_test)
     assert 'accuracy' in score
 
+    # check incumbent config and results
+    incumbent_config, incumbent_results = estimator.get_incumbent_results()
+    assert isinstance(incumbent_config, Configuration)
```
Comment (franchuterivera) [inline test/test_api/test_api.py:184]: Do you think you can create a unit test, like a separate one to test the `include_traditional`?

What we can do is mock.patch the evaluator.run() method to always return a desired value and quickly compare that everything is working as expected?

For example, the tae evaluator run could return:

- time.time() as cost, so we would expect that the best configuration is for sure a lgbm which is the first thing that is ran
- Then without the include, it should be a pytorch mode

What do you think?

File: autoPyTorch/api/base_task.py (hunk 3 above)
Comment (ravinkohli) [reply autoPyTorch/api/base_task.py:652]: so we dont have pipeline configuration in the run info as well as additional info

File: autoPyTorch/api/base_task.py (hunk 4 above)
Comment (ravinkohli) [reply autoPyTorch/api/base_task.py:1255]: yeah agreed

File: autoPyTorch/api/base_task.py (hunk 5 above)
Comment (ravinkohli) [reply autoPyTorch/api/base_task.py:1245]: if we want to be more specific, then yes. I'll do it

File: test/test_api/test_api.py (hunk 6 above)
Comment (ravinkohli) [reply test/test_api/test_api.py:184]: Can you tell me how to just set cost to time.time(). As the TAE.run returns status, cost, runtime, additional_run_info, how can I set cost as time and maintain the other info?

File: autoPyTorch/api/base_task.py (hunk 5 above)
Comment (franchuterivera) [reply autoPyTorch/api/base_task.py]: Thanks, it is not only being specific but if not specified, it is converted to `Dict[Any, Any]` which basically checks almost nothing :)

File: test/test_api/test_api.py (hunk 6 above)
Comment (franchuterivera) [reply test/test_api/test_api.py:185]: Todo a patch please follow [this](https://docs.python.org/3/library/unittest.mock.html)

So what we do is that we set as return value a tuple like:
```
eval_patch.return_value = (StatusType.SUCCESS, time.time(), {})
```

We can also provide a tae function [here](https://github.com/automl/Auto-PyTorch/blob/a4e08e297dc298b53d5962f0f4b4e55b5e9ded8c/autoPyTorch/optimizer/smbo.py#L254)
 I think... maybe it is easier?
